apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsDgsDatasetIdControl.tsx

```diff
@@ -0,0 +1,311 @@
+import type { ControlProps, RankedTester } from "@jsonforms/core"
+import { useEffect, useState } from "react"
+import {
+  Box,
+  FormControl,
+  HStack,
+  Icon,
+  Modal,
+  ModalBody,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  Text,
+  useDisclosure,
+  VStack,
+} from "@chakra-ui/react"
+import { and, isStringControl, rankWith, schemaMatches } from "@jsonforms/core"
+import { withJsonFormsControlProps } from "@jsonforms/react"
+import {
+  Button,
+  FormErrorMessage,
+  FormLabel,
+  Infobox,
+  Input,
+  ModalCloseButton,
+} from "@opengovsg/design-system-react"
+import { DGS_DATASET_ID_FORMAT } from "@opengovsg/isomer-components"
+import { useDebounce } from "@uidotdev/usehooks"
+import { BiLink } from "react-icons/bi"
+import { z } from "zod"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+import { useDgsMetadata } from "~/features/editing-experience/components/form-builder/hooks/useDgsMetadata"
+import { getDgsIdFromString } from "~/features/editing-experience/utils"
+import { useZodForm } from "~/lib/form"
+
+export const jsonFormsDgsDatasetIdControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.TextControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === DGS_DATASET_ID_FORMAT),
+  ),
+)
+
+const generateDgsDatasetUrl = (datasetId: string | null) => {
+  if (!datasetId) return ""
+  return `https://data.gov.sg/datasets/${datasetId}/view`
+}
+
+interface DgsDatasetIdModalProps {
+  isOpen: boolean
+  onClose: () => void
+  onSave: (datasetId: string) => void
+  initialValue?: string
+}
+
+const DgsDatasetIdModal = ({
+  isOpen,
+  onClose,
+  onSave,
+  initialValue = "",
+}: DgsDatasetIdModalProps) => {
+  const initialValueUrl = generateDgsDatasetUrl(initialValue)
+
+  const [inputValue, setInputValue] = useState(initialValueUrl)
+  const debouncedInputValue = useDebounce(inputValue, 300)
+  const isDebouncing = inputValue !== debouncedInputValue
+
+  const datasetId = getDgsIdFromString({ string: debouncedInputValue })
+
+  const { data: format, isLoading: isValidatingDataset } = useDgsMetadata({
+    datasetId,
+  })
+  const isValidDataset = format === "CSV"
+
+  const isLoading = isValidatingDataset || isDebouncing
+
+  const {
+    register,
+    handleSubmit,
+    setError,
+    clearErrors,
+    formState: { errors, isValid },
+  } = useZodForm({
+    mode: "onChange",
+    schema: z.object({
+      datasetId: z
+        .string()
+        .min(1, "Dataset URL is required")
+        .refine((value) => getDgsIdFromString({ string: value }) !== null, {
+          message:
+            "This doesn't look like a valid link from data.gov.sg. Check that you have the correct link and try again.",
+        }),
+    }),
+    reValidateMode: "onChange",
+  })
+
+  // Update inputValue when initialValue changes (when modal opens)
+  useEffect(() => {
+    setInputValue(initialValueUrl)
+  }, [initialValueUrl])
```

Comment:
issue: I think we don't need this `useEffect`? We can force `DgsDatasetIdModal` to re-render below inside `JsonFormsDgsDatasetIdControl` by doing something like this:

```ts
      {isDgsModalOpen && (<DgsDatasetIdModal
        isOpen={isDgsModalOpen}
        onClose={onDgsModalClose}
        onSave={(datasetId) => handleDatasetIdSave(datasetId)}
        initialValue={data || ""}
      />)}
```

This forces a re-render without the need for useEffect, same perf impact anyway since useEffect re-renders the entire component.